com.microsoft.mrtk.input/Visualizers/ControllerVisualizer/ControllerModelLoader.cs

```diff
@@ -0,0 +1,87 @@
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// A helper class which loads and caches controller models fetched from the controller's platform sdk
+    /// </summary>
+    public class ControllerModelLoader
+    {
+        /// <summary>
+        /// A dictionary which caches the controller model gameobject associated with a specified model key
+        /// </summary>
+        public static Dictionary<ulong, GameObject> ControllerModelDictionary = new Dictionary<ulong, GameObject>();
+
+        /// <summary>
+        /// Tries to load the controller model game object from the provided OpenXR ControllerModel.
+        /// Requires the MR OpenXR plugin to work.
+        /// </summary>
+        /// <param name="controllerModelProvider">The OpenXR ControllerModel to loade from</param>
+        /// <returns>A gameobject representing the generated controller model in the scene</returns>
+        public async static Task<GameObject> TryGenerateControllerModelFromPlatformSDK(Handedness handedness)
+        {
+            GameObject gltfGameObject = null;
+
+#if MROPENXR_PRESENT && (UNITY_STANDALONE_WIN || UNITY_WSA || UNITY_ANDROID) && GLTFAST_PRESENT && KTX_PRESENT
+            ControllerModel controllerModelProvider = handedness == Handedness.Left ? ControllerModel.Left :
+                                                        handedness == Handedness.Right ?  ControllerModel.Right :
+                                                        null;
+
+            if (controllerModelProvider == null)
+            {
+                Debug.LogWarning("Controller model provider does not exist for this handedness");
+                return null;
+            }
+
+            if (!controllerModelProvider.TryGetControllerModelKey(out ulong modelKey))
+            {
+                Debug.LogWarning("Failed to obtain controller model key from platform.");
+                return null;
+            }
+
+            if (ControllerModelDictionary.TryGetValue(modelKey, out gltfGameObject))
+            {
+                gltfGameObject.SetActive(true);
+                return gltfGameObject;
+            }
+
+            byte[] modelStream = await controllerModelProvider.TryGetControllerModel(modelKey);
+
+            if (modelStream == null || modelStream.Length == 0)
+            {
+                Debug.LogError("Failed to obtain controller model from platform.");
+                return null;
+            }
+
+            GltfImport gltf = new GltfImport();
+            bool success = await gltf.LoadGltfBinary(modelStream);
+            gltfGameObject = new GameObject(modelKey.ToString());
+            if (success && gltf.InstantiateMainScene(gltfGameObject.transform))
+            {
+                // After all the awaits, double check that another task didn't finish earlier
+                if (ControllerModelDictionary.TryGetValue(modelKey, out GameObject existingGameObject))
```

Comment:
Do you anticipate this to be something that happens often? If so it might be worth pursuing a path where we don't `LoadGltfBinary` multiple times.

Reply:
This class caches existing models which have been successfully loaded so it should be called a minimal number of times. This is more or less the same gltf loading code we had in mrtk2